rfcs/text/0002_encrypted_attributes.md

```diff
@@ -0,0 +1,228 @@
+or `bulkCreate` method and if they try to do so the error will be thrown.
+
+To reduce the risk of unintentional decryption and consequent leaking of the sensitive information there is only one way
+to retrieve saved object and decrypt its encrypted attributes and it's exposed only through `encrypted_saved_objects` plugin:
+
+```typescript
+const savedObject = await server.plugins.encrypted_saved_objects.getDecryptedAsInternalUser(
+  'server-action',
+  'dd9750b9-ef0a-444c-8405-4dfcc2e9d670'
+);
+
+// savedObject = { 
+//   id: 'dd9750b9-ef0a-444c-8405-4dfcc2e9d670',
+//   type: 'server-action',
+//   name: 'my-server-action',
+//   credentials: { username: 'some-user', password: 'some-password' },
+//   apiKey: 'dGhpcyBpcyBub3QgYSByZWFsIHRva2VuIGJ1dCBpdCBpcyBvb',
+// };
+```
+
+As can be seen from the method name, the request to retrieve saved object and decrypt its attributes is performed on
+behalf of the internal Kibana user and hence isn't supposed to be called within user request context.
+
+**Note:** the fact that saved object with encrypted attributes is created using standard Saved Objects API within a 
+particular user and space context, but retrieved out of any context makes it unclear how consumers are supposed to 
+provide that context and retrieve saved object from a particular space. Current plan for `getDecryptedAsInternalUser` 
+method is to accept a third `BaseOptions` argument that allows consumers to specify `namespace` that they can retrieve
+from the request using public `spaces` plugin API.
+
+## Encryption and decryption
+
+Saved object attributes are encrypted using [@elastic/node-crypto](https://github.com/elastic/node-crypto) library. Please
+take a look at the source code of this library to know how encryption is performed exactly, what algorithm and encryption 
+parameters are used, but in short it's AES Encryption with AES-256-GCM that uses random initialization vector and salt.
+
+As with encryption key for Kibana's session cookie, master encryption key used by `encrypted_saved_objects` plugin can be
+defined as a configuration value (`xpack.encrypted_saved_objects.encryptionKey`) via `kibana.yml`, but it's **highly 
+recommended** to define this key in the [Kibana Keystore](https://www.elastic.co/guide/en/kibana/current/secure-settings.html)
+instead. The master key should be cryptographically safe and be equal or greater than 32 bytes.
+
+To prevent certain vectors of attacks where raw content of encrypted attributes of one saved object is copied to another
+saved object that would unconsciously allow to decrypt content that was not supposed to be decrypted we rely on Additional
+authenticated data (AAD) during encryption and decryption. AAD consists of the following components:
+
+* Saved object ID
+* Saved object type
+* Saved object attributes
+ 
+AAD does not include encrypted attributes themselves and attributes defined in optional `attributesToExcludeFromAAD` 
```

Comment:
As I mentioned above, it would be good to explain _why_ you'd want to do this.

Reply:
Sure, I'll add explanation (e.g. field can contain quite "heavy" data that, if involved into encryption/AAD generation, may slow down encryption/decryption significantly especially during bulk operations).